iwyu.cc

```diff
@@ -2676,19 +2672,33 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
       const ASTNode* ast_node = MostElaboratedAncestor(current_ast_node());
       if (!ast_node->ParentIsA<TypedefNameDecl>()) {
         const TypedefNameDecl* typedef_decl = typedef_type->getDecl();
+        const Type* type = typedef_decl->getUnderlyingType().getTypePtr();
         const set<const Type*>& provided_with_typedef =
-            GetProvidedTypesForTypedef(typedef_decl);
+            GetProvidedTypesForAlias(type, GetLocation(typedef_decl));
         InsertAllInto(provided_with_typedef, &blocked_types);
         VERRS(6) << "User, not author, of typedef "
                  << typedef_decl->getQualifiedNameAsString()
                  << " owns the underlying type:\n";
         // If any of the used types are themselves typedefs, this will
         // result in a recursive expansion.
-        const Type* type = typedef_decl->getUnderlyingType().getTypePtr();
         ReportTypeUseInternal(used_loc, type, blocked_types, deref_kind);
       }
       return;
     }
+    if (const auto* template_spec_type =
```

Comment:
I see the typedef case on line 2671/2667 is using `getAs<>` and we're using `dyn_cast<>` here.

I can't fully understand which is the best here (`getAs<>` desugars and `dyn_cast<>` does not, as I understand it)... But maybe they should use the same mechanism? Or is it important that they don't?

Reply:
If it is `TemplateSpecializationType`, it should be exactly `TemplateSpecializationType` after desugaring at the beginning of `ReportTypeUseInternal`. But, yeah, let the code be similar.